service/history/replicationTaskProcessor.go

```diff
@@ -0,0 +1,291 @@
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"time"
+
+	h "github.com/uber/cadence/.gen/go/history"
+	r "github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/service/worker/replicator"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+const (
+	dropSyncShardTaskTimeThreshold = 10 * time.Minute
+	replicationTimeout             = 30 * time.Second
+)
+
+var (
+	// ErrUnknownReplicationTask is the error to indicate unknown replication task type
+	ErrUnknownReplicationTask = &shared.BadRequestError{Message: "unknown replication task"}
+)
+
+type (
+	replicationTaskProcessor struct {
+		shard            ShardContext
+		readLevel        int64
+		historyEngine    Engine
+		sourceCluster    string
+		domainReplicator replicator.DomainReplicator
+		metricsClient    metrics.Client
+		logger           log.Logger
+		retryPolicy      backoff.RetryPolicy
+
+		requestChan chan<- *request
+		done        chan struct{}
+	}
+
+	request struct {
+		token    *r.ReplicationToken
+		respChan chan<- *r.ReplicationTasksInfo
+	}
+)
+
+func newReplicationTaskProcessor(
+	shard ShardContext,
+	historyEngine Engine,
+	domainReplicator replicator.DomainReplicator,
+	metricsClient metrics.Client,
+	replicationTaskFetcher *replicationTaskFetcher,
+) *replicationTaskProcessor {
+	retryPolicy := backoff.NewExponentialRetryPolicy(time.Second)
+	retryPolicy.SetBackoffCoefficient(1)
+	retryPolicy.SetMaximumAttempts(5)
+
+	return &replicationTaskProcessor{
+		shard:            shard,
+		historyEngine:    historyEngine,
+		sourceCluster:    replicationTaskFetcher.GetSourceCluster(),
+		domainReplicator: domainReplicator,
+		metricsClient:    metricsClient,
+		logger:           shard.GetLogger(),
+		retryPolicy:      retryPolicy,
+		requestChan:      replicationTaskFetcher.GetRequestChan(),
+		done:             make(chan struct{}),
+	}
+}
+
+func (p *replicationTaskProcessor) Start() {
+	go func() {
+		p.readLevel = p.shard.GetClusterReplicationLevel(p.sourceCluster)
+		scope := p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster))
+
+		for {
+			respChan := make(chan *r.ReplicationTasksInfo, 1)
+			p.requestChan <- &request{
+				token:    &r.ReplicationToken{ShardID: common.Int32Ptr(int32(p.shard.GetShardID())), TaskID: common.Int64Ptr(p.readLevel)},
+				respChan: respChan,
+			}
+
+			select {
+			case response := <-respChan:
+				p.logger.Debug("Got fetch replication tasks response.",
+					tag.ReadLevel(response.GetReadLevel()),
+					tag.Bool(response.GetHasMore()),
+					tag.Counter(len(response.GetReplicationTasks())),
+				)
+
+				for _, replicationTask := range response.ReplicationTasks {
+					p.processTask(replicationTask)
```

Comment:
no error?

Reply:
error handling happens inside processTask. Also there is a todo for DLQ, which should also happen inside. So we don't have error here.